distributed/scheduler.py

```diff
@@ -5205,6 +5207,10 @@ def remove_plugin(self, plugin):
         """Remove external plugin from scheduler"""
         self.plugins.remove(plugin)
 
+    def register_scheduler_plugin(self, comm=None, plugin=None):
+        """TODO docstring."""
+        self.add_plugin(plugin=loads(plugin))
```

Comment:
We should add some additional logic here such that, similar to `WorkerPlugin.setup`, `SchedulerPlugin.start` also runs when a new plugin is dynamically added to the scheduler through `Client.register_scheduler_plugin`. FWIW that's one reason why `test_register_scheduler_plugin` isn't currently passing

Reply:
I'm not exactly following here. What part of the worker plugin registration API is connected to `WorkerPlugin.setup`? Does the `_WorkerSetupPlugin` class down at https://github.com/dask/distributed/blob/05c56214cca6bccb0de0cbf85d7f01990f6923d8/distributed/client.py#L4129
have something to due with you're mentioning here?

Reply:
Today the `setup` method for `WorkerPlugin`s are [run in two scenerios](https://distributed.dask.org/en/latest/plugins.html#distributed.diagnostics.plugin.WorkerPlugin.setup):

1. When a new worker starts up, the `setup` method for all currently registered `WorkerPlugin`s will be run
2. When a new `WorkerPlugin` is registered, the corresponding `setup` method is run on all existing workers

https://github.com/dask/distributed/blob/05c56214cca6bccb0de0cbf85d7f01990f6923d8/distributed/worker.py#L2796-L2804

For `SchedulerPlugin`s we want to have a similar behavior. Since there's only ever a single scheduler, we don't need to worry about the first case above because we won't ever have new schedulers arriving in the cluster. But we do want to make sure that, like the second case above, when a new `SchedulerPlugin` is registered, that the corresponding `start` method (there's a slight discrepancy in the `start` / `setup` method name used between scheduler and worker plugins) is run on the scheduler. Does that help clarify things?

Reply:
To be clear, the reason we want to make sure `WorkerPlugin.setup` / `Scheduler.start` are run when a new plugin is registered is because often times those `setup`/`start` methods contain setup code that must be run before other plugin methods (e.g. `add_worker` or `transition`) will work

Reply:
Ah yes I see thanks for the detailed explanation! OK I just added the logic to call `plugin.start(self)` in `Scheduler.register_scheduler_plugin`. I'm still trying to figure out how best to test this, but do let me know if the `start` logic is going in the right direction.

Reply:
After rereading the commit, should the logic be in `Scheduler.add_plugin` over `Scheduler.register_scheduler_plugin`?

Reply:
That might break existing usage where, presumably, the caller had to ensure `start()` happened. I don't know if there are any of these in practice; or indeed how you could tell whether `start` has been called. It probably should be called immediately after instantiation (or deserialisation).

Reply:
>  I'm still trying to figure out how best to test this, but do let me know if the start logic is going in the right direction.

Yeah, your updates look great 👍 The `test_register_scheduler_plugin` test you've added should already ensure that `Client.register_scheduler_plugin` results in `SchedulerPlugin.start` being called.

> That might break existing usage where, presumably, the caller had to ensure start() happened

Yep, agreed that could break some existing setups where users previously had to call `SchedulerPlugin.start` manually